woocommerce/woocommerce-ios
# Recover from date decoding failures when a fallback or optional date can be used

Closes: #11486


## Why
A number of networking entities can fail to decode if they have a date field that can't be decoded as a `Date`.

This caused ~3,500 JSON decoding errors so far this month, mostly with these date fields:

* `date_created_gmt`
* `date_modified_gmt`

## How

This fix allows entities to recover from these date decoding failures by allowing a `nil` date (for optional properties) or using a fallback if the date can't be decoded. In many cases the fallback already existed but was only used if the date field was missing, not if the decoding failed.

The fix is applied to the following entities:

* Coupon
* Media
* WordPressMedia
* Order
* OrderNote
* Product
* ProductImage
* ProductReview
* ProductVariation
* Refund

Note: This fix doesn't extend to all date properties, even in these entities. Using fallbacks for some dates that can be updated from the app (e.g. sale dates for `Product`) could overwrite remote data that affects the store in unexpected and potentially negative ways.

## Testing instructions

Optional testing:

1. Prerequisite: Use a tool like Charles Proxy to modify the responses with an unexpected date format (any unexpected format should be fine; see example for images in the `products` response in the original issue: https://github.com/woocommerce/woocommerce-ios/issues/11486).
2. Build and run the app.
3. Open the order tab and confirm the order list loads.
4. Select an order and confirm order notes and refunds load.
5. Open the products tab and confirm the product list loads.
6. Select a product and confirm the product images, reviews, and any variation details load.
7. Open the menu tab > Coupons and confirm the coupon list loads.


---
